### PR TITLE
reference/best-practices: add HAProxy best practices

### PR DESCRIPTION
**Pull request by superlzs0476**
### What is changed, added or deleted? 
Add haproxy best practices for TiDB  Documentation

Provide load balancing solutions for TiDB users.

### CODE REVIEW

File: v3.0/reference/best-practices/Haproxy-best-practices.md [hunk 1]
```diff
@@ -0,0 +1,192 @@
+---
```
Comment (lilin90) [inline v3.0/reference/best-practices/Haproxy-best-practices.md]: Please use lower-case letters for the file name.

File: v3.0/reference/best-practices/Haproxy-best-practices.md [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+|-sf/-st|\<unix_socket> get listening sockets from a unix socket|在启动后，在pidlist中发送FINISH信号给pid。收到此信号的进程将等待所有会话在退出之前完成。此选项必须最后指定，后跟任意数量的PID。 从技术上讲，SIGTTOU和SIGUSR1都被发送。|
+|-x|\<unix_socket>[,<bind options>...] new stats socket for the master|获取 socket 信息|
+|-S|\<unix_socket>[,\<bind options>...] new stats socket for the master|分配新的 socket|
+
+### 配置 HAProxy
+
+yum 安装过程中会生成配置模版。
+
+```yaml
+global  # 全局配置
```
Comment (jackysp) [inline v3.0/reference/best-practices/Haproxy-best-practices.md]: Please align the comments.

File: v3.0/reference/best-practices/haproxy-best-practices.md [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+HAProxy 配置 Database 负载均衡场景操作简单，以下部署操作具有普遍性，不具有特殊性，建议根据实际场景，个性化配置相关的[配置文件](http://cbonte.github.io/haproxy-dconv/1.9/configuration.html)。
+
+> **注意：**
+>
+> 根据官方建议，目前稳定版本的 HAProxy 为稳定版 [2.0 特性](https://www.haproxy.com/blog/haproxy-2-0-and-beyond/)。
+
+![HAProxy](/media/Haproxy.jpg)
+
+### 安装 HAproxy
```
Comment (jackysp) [inline v3.0/reference/best-practices/haproxy-best-practices.md]:
```suggestion
### 安装 HAProxy
```

**Comment by lilin90** [comment]
@superlzs0476 Please address comments. Thanks!

File: v3.0/reference/best-practices/haproxy-best-practices.md [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+Usage : haproxy [-f <cfgfile|cfgdir>]* [ -vdVD ] [ -n <maxconn> ] [ -N <maxpconn> ]
+        [ -p <pidfile> ] [ -m <max megs> ] [ -C <dir> ] [-- <cfgfile>*]
+```
+
+|参数|描述|详情|
+|:-----|:---|:---|
+|-v|displays version|版本|
+|-vv|shows known build options.|版本|
+|-d|enters debug mode|debug 模式开启|
+| -db|only disables background mode.|禁止后台模式|
```
Comment (superlzs0476) [inline v3.0/reference/best-practices/haproxy-best-practices.md]:
```suggestion
| -db|only disables background mode.|仅禁止后台模式|
```

File: v3.0/reference/best-practices/haproxy-best-practices.md [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+   systemctl start haproxy.service
+   ```
+
+### systemd 停止 HAProxy
+
+- 方法一：执行 `kill -9`
+
+   ```bash
+   ps -ef |grep haproxy
+   kill -9 ${haproxy.pid}  # HAProxy 进程 pid
```
[the remark this answers is not in the document]
Comment (superlzs0476) [reply v3.0/reference/best-practices/haproxy-best-practices.md]:
```suggestion
kill -9 ${haproxy.pid}  # 终止 HAProxy 相关的 pid 进程
```

File: v3.0/reference/best-practices/haproxy-best-practices.md [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+|-N|sets the default, per-proxy maximum # of connections (2000)|单点最大连接数|
+|-L|set local peer name (default to hostname)|本地实例对等名称|
+|-p|writes pids of all children to this file|haproxy 进程 pid 信息写入 file |
+|-de|disables epoll() usage even when available|禁止使用speculative epoll。epoll仅在Linux 2.6和某些定制的Linux 2.4系统上可用。|
+|-dp|disables poll() usage even when available|禁止使用epoll。epoll仅在Linux 2.6和某些定制的Linux 2.4系统上可用。|
+|-dS|disables splice usage (broken on old kernels)|禁止使用 speculative epoll，epoll 仅在 Linux 2.6 和某些定制的 Linux 2.4 系统上可用。|
+|-dR|disables SO_REUSEPORT usage|禁止使用 SO_REUSEPORT|
+|-dr|ignores server address resolution failures|忽略失败的 server|
+|-dV|disables SSL verify on servers side|禁止在 servers 端使用 SSL|
+|-sf/-st|\<unix_socket> get listening sockets from a unix socket|在启动后，在pidlist中发送FINISH信号给pid。收到此信号的进程将等待所有会话在退出之前完成。此选项必须最后指定，后跟任意数量的PID。 从技术上讲，SIGTTOU和SIGUSR1都被发送。|
```
Comment (superlzs0476) [inline v3.0/reference/best-practices/haproxy-best-practices.md]:
```suggestion
|-sf/-st|\<unix_socket> get listening sockets from a Unix socket|在启动后，在 pidlist中发送 FINISH 信号给 pid 。收到此信号的进程将等待所有会话在退出之前完成，即优雅停止服务。此选项必须最后指定，后跟任意数量的 pid，SIGTTOU 和 SIGUSR1 都被发送。|
```

File: v3.0/reference/best-practices/haproxy-best-practices.md [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+|-W|master-worker mode.|主从模式|
+|-q|quiet mode : don't display messages|静默模式，不输出信息|
+|-c|check mode : only check config files and exit|检查配置信息文件|
+|-n|sets the maximum total # of connections (2000)|最大总连接数|
+|-m|limits the usable amount of memory (in MB)|最大使用内存|
+|-N|sets the default, per-proxy maximum # of connections (2000)|单点最大连接数|
+|-L|set local peer name (default to hostname)|本地实例对等名称|
+|-p|writes pids of all children to this file|haproxy 进程 pid 信息写入 file |
+|-de|disables epoll() usage even when available|禁止使用speculative epoll。epoll仅在Linux 2.6和某些定制的Linux 2.4系统上可用。|
+|-dp|disables poll() usage even when available|禁止使用epoll。epoll仅在Linux 2.6和某些定制的Linux 2.4系统上可用。|
```
Comment (superlzs0476) [inline v3.0/reference/best-practices/haproxy-best-practices.md]:
```suggestion
|-dp|disables poll() usage even when available|禁止使用 epoll。epoll 仅在 Linux 2.6 和某些定制的 Linux 2.4 系统上可用。|
```

File: v3.0/reference/best-practices/haproxy-best-practices.md [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+Usage : haproxy [-f <cfgfile|cfgdir>]* [ -vdVD ] [ -n <maxconn> ] [ -N <maxpconn> ]
+        [ -p <pidfile> ] [ -m <max megs> ] [ -C <dir> ] [-- <cfgfile>*]
+```
+
+|参数|描述|详情|
+|:-----|:---|:---|
+|-v|displays version|版本|
+|-vv|shows known build options.|版本|
+|-d|enters debug mode|debug 模式开启|
+| -db|only disables background mode.|禁止后台模式|
```
[the remark this answers is not in the document]
Comment (superlzs0476) [reply v3.0/reference/best-practices/haproxy-best-practices.md]:
```suggestion
| -db|only disables background mode.|仅禁止后台模式|
```

File: v3.0/reference/best-practices/haproxy-best-practices.md (hunk 7 above)
Comment (superlzs0476) [inline v3.0/reference/best-practices/haproxy-best-practices.md]:
```suggestion
|-dp|disables poll() usage even when available|禁止使用 epoll, epoll 仅在 Linux 2.6 和某些定制的 Linux 2.4 系统上可用。|
```

File: v3.0/reference/best-practices/haproxy-best-practices.md [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+
+### 配置 HAProxy
+
+yum 安装过程中会生成配置模版。
+
+```yaml
+global                                     # 全局配置
+   log         127.0.0.1 local2            # 定义全局的 syslog 服务器，最多可以定义两个
+   chroot      /var/lib/haproxy            # 将当前目录为指定目录，设置超级用户权限启动进程，提高安全性
+   pidfile     /var/run/haproxy.pid        # 将 haproxy 进程写入 pid 文件
```
Comment (superlzs0476) [inline v3.0/reference/best-practices/haproxy-best-practices.md]:
```suggestion
   pidfile     /var/run/haproxy.pid        # 将 HAproxy 进程写入 pid 文件
```

Comment (superlzs0476) [inline v3.0/reference/best-practices/haproxy-best-practices.md]:
```suggestion
   pidfile     /var/run/haproxy.pid        # 将 HAproxy 进程写入 pid 文件
```

File: v3.0/reference/best-practices/haproxy-best-practices.md [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+   maxconn     4000                        # 设置每个 haproxy 进程锁接受的最大并发连接数
```
Comment (superlzs0476) [inline v3.0/reference/best-practices/haproxy-best-practices.md]:
```suggestion
   maxconn     4000                        # 设置每个 HAproxy 进程锁接受的最大并发连接数
```

File: v3.0/reference/best-practices/haproxy-best-practices.md [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+   group       haproxy                     # 同 gid 参数，建议专用用户组
+   nbproc      40                          # 启动多个进程来转发请求，需要调整到足够大的值来保证 haproxy 本身不会成为瓶颈
+   daemon                                  # 让 HAProxy 以守护进程的方式工作于后台，等同于“-D”选项的功能。当然，也可以在命令行中用“-db”选项将其禁用。
+   stats socket /var/lib/haproxy/stats     # 定义统计信息保存位置
+
+defaults                                   # 默认配置
+   log global                              # 日志继承全局配置段的设置
+   retries 2                               # 向上游服务器尝试连接的最大次数，超过此值就认为后端服务器不可用
+   timeout connect  2s                     # HAProxy 与后端服务器连接超时时间，如果在同一个局域网内可设置成较短的时间
+   timeout client 30000s                   # 定义客户端与haproxy连接后，数据传输完毕，不再有数据传输，即非活动连接的超时时间
```
Comment (superlzs0476) [inline v3.0/reference/best-practices/haproxy-best-practices.md]:
```suggestion
   timeout client 30000s                   # 定义客户端与 HAproxy 连接后，数据传输完毕，不再有数据传输，即非活动连接的超时时间
```

File: v3.0/reference/best-practices/haproxy-best-practices.md [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+   timeout server 30000s                   # 定义haproxy与上游服务器非活动连接的超时时间
```
Comment (superlzs0476) [inline v3.0/reference/best-practices/haproxy-best-practices.md]:
```suggestion
   timeout server 30000s                   # 定义 HAproxy 与上游服务器非活动连接的超时时间
```

File: v3.0/reference/best-practices/haproxy-best-practices.md [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+
+# HAProxy 最佳实践
+
+本文介绍 HAProxy 在 TiDB 中的最佳实践。
+
+## HAProxy 简介
+
+[HAProxy](https://github.com/haproxy/haproxy) 是由 C 语言编写的自由开放源码的软件。为基于 TCP 和 HTTP 协议的应用程序提供高可用性、负载均衡和代理服务。它在对 CPU 和 内存的快速及高效使用方面的特点，获得诸如：GitHub、Bitbucket、Stack Overflow、Reddit、Tumblr、Twitter 和 Tuenti 在内的知名网站，及亚马逊网络服务系统的广泛使用。
+
+HAProxy 是由 Linux 内核的核心贡献者 Willy Tarreau 于 2000 年编写，并仍然负责该项目的维护，该在开源社区提供免费和版本迭代。最新的稳定版本 2.0.0 于 2019 年 8 月 16 日发布，并带来更多的[优秀的特性](https://www.haproxy.com/blog/haproxy-2-0-and-beyond/)。
```
[the remark this answers is not in the document]
Comment (superlzs0476) [reply v3.0/reference/best-practices/haproxy-best-practices.md]:
```suggestion
HAProxy 是由 Linux 内核的核心贡献者 Willy Tarreau 于 2000 年编写，他现在仍然负责该项目的维护，并在开源社区免费提供版本迭代。最新的稳定版本 2.0.0 于 2019 年 8 月 16 日发布，带来更多[优秀的特性](https://www.haproxy.com/blog/haproxy-2-0-and-beyond/)。
```

File: v3.0/reference/best-practices/haproxy-best-practices.md [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+   timeout server 30000s                   # 定义haproxy与上游服务器非活动连接的超时时间
+
+listen admin_stats                         # frontend和backend的组合体,监控组的名称，按需自定义名称
+   bind 0.0.0.0:8080                       # 配置监听端口
+   mode http                               # 配置监控运行的模式，在这为http模式
+   option httplog                          # 表示开始打开记录http请求的日志功能
+   maxconn 10                              # 最大并发连接数
+   stats refresh 30s                       # 配置每隔 30 秒自动刷新监控页面
+   stats uri /haproxy                      # 配置监控页面的url
+   stats realm Haproxy                     # 配置监控页面的提示信息
```
[the remark this answers is not in the document]
Comment (superlzs0476) [reply v3.0/reference/best-practices/haproxy-best-practices.md]: HAproxy

File: v3.0/reference/best-practices/haproxy-best-practices.md [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+   systemctl start haproxy.service
+   ```
+
+### 使用 `systemd` 停止 HAProxy
+
+- 方法一：执行 `kill -9`
+
+   ```bash
+   ps -ef | grep haproxy
+   kill -9 ${haproxy.pid}  # 终止 HAProxy 相关的 pid 进程
```
Comment (superlzs0476) [inline v3.0/reference/best-practices/haproxy-best-practices.md]:
```suggestion
   kill -9 ${haproxy.pid}  # 终止 HAproxy 相关的 pid 进程
```

**Pull request closed without merging**